File: backend/app/schemas/kpi.py
from datetime import datetime
from typing import Literal

from app.schemas.common import CamelModel
from app.schemas.lead import LeadListItem, LeadStatus


class KpiSummary(CamelModel):
    leads_today: int
    leads_today_delta: int
    qualified_today: int
    qualified_rate: float
    booked_today: int
    avg_response_time_seconds: float
    avg_response_time_delta: float


class KpiTimeseriesPoint(CamelModel):
    date: str
    new_leads: int
    qualified: int
    booked: int
    rejected: int
    avg_response_time_seconds: float


class DashboardOverview(CamelModel):
    summary: KpiSummary
    timeseries: list[KpiTimeseriesPoint]
    recent_leads: list[LeadListItem]


class SseNewLeadEvent(CamelModel):
    type: Literal["new_lead"] = "new_lead"
    lead: LeadListItem


class SseStatusChangeEvent(CamelModel):
    type: Literal["status_change"] = "status_change"
    lead_id: str
    status: LeadStatus


class SseHeartbeatEvent(CamelModel):
    type: Literal["heartbeat"] = "heartbeat"
    timestamp: datetime

File: frontend/lib/schema/kpi.ts
import { z } from "zod";
import { LeadListItemSchema, LeadStatusSchema } from "./lead";

export const KpiSummarySchema = z.object({
  leadsToday: z.number(),
  leadsTodayDelta: z.number(),
  qualifiedToday: z.number(),
  qualifiedRate: z.number().min(0).max(1),
  bookedToday: z.number(),
  avgResponseTimeSeconds: z.number(),
  avgResponseTimeDelta: z.number(),
});
export type KpiSummary = z.infer<typeof KpiSummarySchema>;

export const KpiTimeseriesPointSchema = z.object({
  date: z.string(),
  newLeads: z.number(),
  qualified: z.number(),
  booked: z.number(),
  rejected: z.number(),
  avgResponseTimeSeconds: z.number(),
});
export type KpiTimeseriesPoint = z.infer<typeof KpiTimeseriesPointSchema>;

export const DashboardOverviewSchema = z.object({
  summary: KpiSummarySchema,
  timeseries: z.array(KpiTimeseriesPointSchema),
  recentLeads: z.array(LeadListItemSchema),
});
export type DashboardOverview = z.infer<typeof DashboardOverviewSchema>;

export const SseEventSchema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("new_lead"), lead: LeadListItemSchema }),
  z.object({ type: z.literal("status_change"), leadId: z.string(), status: LeadStatusSchema }),
  z.object({ type: z.literal("heartbeat"), timestamp: z.string() }),
]);
export type SseEvent = z.infer<typeof SseEventSchema>;
